### PR TITLE
Improve generation

### PR DESCRIPTION
**Pull request by gejinchen**
# What does this PR do?

Some changes to improve text generation so that it would be easier to integrate the run_generation.py task.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/generation_utils.py [hunk 1]
```diff
@@ -1387,11 +1385,17 @@ def greedy_search(
...
         unfinished_sequences = input_ids.new(input_ids.shape[0]).fill_(1)
-        input_ids = self._pad_input_ids_to_max_length(input_ids, max_length)
-
-        cur_len = torch.argmax((input_ids == self.config.pad_token_id).to(torch.long), dim=1)[0]
+        cur_len = input_ids.shape[-1]
 
         while True:
+
+            input_ids = self._pad_tensors_to_max_len(input_ids, max_length, pad_token_id)
+            # For a seq2seq model such as BART, the "attention_mask" is the enocder/cross attention mask and it does not require padding.
```
Comment (michaelbenayoun) [inline optimum/graphcore/generation_utils.py:1393]: Typoe enocder -> encoder

File: optimum/graphcore/generation_utils.py [hunk 2]
```diff
@@ -1404,7 +1408,16 @@ def greedy_search(
...
             outputs = self._poptorch_outputs_to_model_outputs(outputs)
 
-            next_token_logits = outputs["logits"][:, cur_len - 1, :].to(torch.float32)
+            # Remove padding and restore to actual length
+            input_ids = input_ids[:, :cur_len]
+            if not self.config.is_encoder_decoder:
+                model_kwargs["attention_mask"] = model_kwargs["attention_mask"][:, :cur_len]
+
+            if outputs.logits.dim() == 3:
+                next_token_logits = outputs.logits[:, cur_len - 1, :]
```
Comment (michaelbenayoun) [inline optimum/graphcore/generation_utils.py:1417]: Should it still be `cur_len - 1`?

Comment (gejinchen) [reply optimum/graphcore/generation_utils.py:1417]: For BART and T5, I didn't use gather to pick out the logits at the last position because it is actually more complicated than GPT2, and also because they only need 128 length anyway. That is why I use this if/else statement to differentiate the gather/no gather situations. Dimension is 3 means no gather, so the index should still be `cur_len - 1`.

File: optimum/graphcore/generation_utils.py (hunk 1 above)
Comment (gejinchen) [reply optimum/graphcore/generation_utils.py:1393]: Thanks!
